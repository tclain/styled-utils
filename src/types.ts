export type HtmlTags =
  | 'a'
  | 'abbr'
  | 'address'
  | 'area'
  | 'article'
  | 'aside'
  | 'audio'
  | 'b'
  | 'base'
  | 'bdi'
  | 'bdo'
  | 'big'
  | 'blockquote'
  | 'body'
  | 'br'
  | 'button'
  | 'canvas'
  | 'caption'
  | 'cite'
  | 'code'
  | 'col'
  | 'colgroup'
  | 'data'
  | 'datalist'
  | 'dd'
  | 'del'
  | 'details'
  | 'dfn'
  | 'dialog'
  | 'div'
  | 'dl'
  | 'dt'
  | 'em'
  | 'embed'
  | 'fieldset'
  | 'figcaption'
  | 'figure'
  | 'footer'
  | 'form'
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'head'
  | 'header'
  | 'hgroup'
  | 'hr'
  | 'html'
  | 'i'
  | 'iframe'
  | 'img'
  | 'input'
  | 'ins'
  | 'kbd'
  | 'keygen'
  | 'label'
  | 'legend'
  | 'li'
  | 'link'
  | 'main'
  | 'map'
  | 'mark'
  | 'marquee'
  | 'menu'
  | 'menuitem'
  | 'meta'
  | 'meter'
  | 'nav'
  | 'noscript'
  | 'object'
  | 'ol'
  | 'optgroup'
  | 'option'
  | 'output'
  | 'p'
  | 'param'
  | 'picture'
  | 'pre'
  | 'progress'
  | 'q'
  | 'rp'
  | 'rt'
  | 'ruby'
  | 's'
  | 'samp'
  | 'script'
  | 'section'
  | 'select'
  | 'small'
  | 'source'
  | 'span'
  | 'strong'
  | 'style'
  | 'sub'
  | 'summary'
  | 'sup'
  | 'table'
  | 'tbody'
  | 'td'
  | 'textarea'
  | 'tfoot'
  | 'th'
  | 'thead'
  | 'time'
  | 'title'
  | 'tr'
  | 'track'
  | 'u'
  | 'ul'
  | 'var'
  | 'video'
  | 'wbr'

  // SVG
  | 'circle'
  | 'clipPath'
  | 'defs'
  | 'ellipse'
  | 'foreignObject'
  | 'g'
  | 'image'
  | 'line'
  | 'linearGradient'
  | 'mask'
  | 'path'
  | 'pattern'
  | 'polygon'
  | 'polyline'
  | 'radialGradient'
  | 'rect'
  | 'stop'
  | 'svg'
  | 'text'
  | 'tspan'

/** A theme| basically a one level object */
export interface ThemeLike {
  [key: string]: string | number
}

export interface PropsWithTheme {
  theme: ThemeLike
}

/** A restricted key from a theme */
export type ThemeLikeKey<ThemeLikeType extends ThemeLike> = keyof ThemeLikeType

/** A function that can take the props  and output a key in the theme */
export type PropsToTheme<Props extends {}, Theme extends ThemeLike> = (
  props: Props
) => ThemeLikeKey<Theme> | string

/** A new type for an interpolation, can be either a theme key of a function that take props and return a theme key */
export type ThemedInterpolation<
  Props extends PropsWithTheme = { theme: {} },
  Theme extends ThemeLike = {}
> = PropsToTheme<Props, Theme> | ThemeLikeKey<Theme>
