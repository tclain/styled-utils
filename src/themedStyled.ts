import { ThemedInterpolation, ThemeLike, PropsWithTheme, HtmlTags } from './types'

/** Now our interpolation function can return theme keys so we need the resolve the actual value from the props */
function resolveInterpolation(
  interpolation: ThemedInterpolation | string,
  props: PropsWithTheme
): unknown {
  const { theme } = props
  if (typeof interpolation === 'function') {
    // get the result of the interpolation
    const possibleThemeKey = interpolation(props)
    return resolveInterpolation(possibleThemeKey, props)
  }
  if (typeof interpolation === 'string') {
    // the result can be a key in theme object, so try to fetch it
    const possibleThemeValue = theme[interpolation]
    // if the theme key exists return it
    if (possibleThemeValue !== undefined) return possibleThemeValue
    // else return the value directly (string, css`` call...)
    return interpolation
  }
  return ''
}

/** a tagged template with themed superpowers
 * It can return a key of a theme in an autocompleted manner thanks to typescript
 */
export function createThemedStyled<Theme extends ThemeLike = any>(styledLikeFunction: Function) {
  return function(tagName: HtmlTags) {
    return function<Props extends {}>(
      strings: TemplateStringsArray,
      ...interpolations: ThemedInterpolation<Props & { theme: Theme }, Theme>[]
    ) {
      const resolvedInterpolations = interpolations.map(interpolation => (props: PropsWithTheme) =>
        resolveInterpolation(interpolation as ThemedInterpolation, props)
      )
      return styledLikeFunction(tagName)(strings, ...resolvedInterpolations)
    }
  }
}
