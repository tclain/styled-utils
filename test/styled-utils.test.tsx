import * as React from 'react'
import emotionStyled from '@emotion/styled'
import { createThemedStyled } from '../src/styled-utils'

import * as renderer from 'react-test-renderer'
import { matchers } from 'jest-emotion'

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers)

const theme = {
  'bg-color': 'blue',
  color: 'red',
  otherColor: 'green'
}

/** a new styled with superpowers */
const styled = createThemedStyled<typeof theme>(emotionStyled)

/**
 * Dummy test
 */
describe('createThemedStyled', () => {
  it('should support theme keys along ', () => {
    const Button = styled('div')<{ primary: boolean }>`
      background-color: ${'bg-color'};
      ${[] as any}
    `
    const tree = renderer.create(<Button />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('it should still support interpolation function', () => {
    const Button = emotionStyled('div')<{ primary?: boolean }>`
      color: ${props => (props.primary ? theme.color : 'teal')};
      /** direct string */
      border-color: ${'teal' as any};
    `
    const tree = renderer.create(<Button primary />).toJSON()
    expect(tree).toMatchSnapshot('primary')

    const notprimary = renderer.create(<Button />).toJSON()
    expect(notprimary).toMatchSnapshot('notprimary')
  })
})
