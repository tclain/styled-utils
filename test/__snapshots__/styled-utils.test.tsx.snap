// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createThemedStyled it should still support interpolation function: notprimary 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-1mmc512{color:teal;border-color:teal;}",
      }
    }
    data-emotion-css="1mmc512"
  />,
  <div
    className="css-1mmc512"
  />,
]
`;

exports[`createThemedStyled it should still support interpolation function: primary 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-9qoy8n{color:red;border-color:teal;}",
      }
    }
    data-emotion-css="9qoy8n"
  />,
  <div
    className="css-9qoy8n"
  />,
]
`;

exports[`createThemedStyled should support theme keys along  1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-t31ymc{background-color:bg-color;}",
      }
    }
    data-emotion-css="t31ymc"
  />,
  <div
    className="css-t31ymc"
  />,
]
`;
